openedx/core/djangoapps/content/course_structures/api/v0/api.py

```diff
@@ -60,15 +67,41 @@ def course_structure(course_key):
                 blocks.
     Raises:
         CourseStructureNotAvailableError, CourseNotFoundError
+
     """
     course = _retrieve_course(course_key)
-    try:
-        requested_course_structure = models.CourseStructure.objects.get(course_id=course.id)
-        return serializers.CourseStructureSerializer(requested_course_structure.structure).data
-    except models.CourseStructure.DoesNotExist:
-        # If we don't have data stored, generate it and return an error.
-        tasks.update_course_structure.delay(unicode(course_key))
-        raise CourseStructureNotAvailableError
+
+    modified_timestamp = models.CourseStructure.objects.filter(course_id=course_key).values('modified')
+    if modified_timestamp.exists():
+        cache_key = 'openedx.content.course_structures.api.v0.api.course_structure.{}.{}.{}'.format(
+            course_key, modified_timestamp[0]['modified'], '_'.join(block_types or [])
+        )
+        data = cache.get(cache_key)  # pylint: disable=maybe-no-member
+        if data is not None:
+            return data
+
+        try:
+            requested_course_structure = models.CourseStructure.objects.get(course_id=course.id)
+        except models.CourseStructure.DoesNotExist:
+            pass
+        else:
+            structure = requested_course_structure.structure
+            if block_types is not None:
+                blocks = requested_course_structure.ordered_blocks
+                required_blocks = OrderedDict()
+                for usage_id, block_data in blocks.iteritems():
+                    if block_data['block_type'] in block_types:
+                        required_blocks[usage_id] = block_data
+
+                structure['blocks'] = required_blocks
+
+            data = CourseStructureSerializer(structure).data
+            cache.set(cache_key, data, None)  # pylint: disable=maybe-no-member
```

Comment:
@e0d A few more info about the cache item is here:

1.  A new item will be created every time this api is called after course publish event.
2.  The item will consist of a list of dicts about all or some of the blocks in the course.
3.  There is no expiry for the item that we put in the cache.

Are there any scaling concerns around this ?